### PR TITLE
V2.1.0-beta.1: Tune the CPU/Memory usage of Client/Session Processes

### PR DESCRIPTION
**Pull request by emqplus**
1. Use gen_server2 and hibernate to improve the CPU/Memory usage of Client/Session Processes
2. The broker requires R19 to build since 2.1 release for emqttd_session uses the maps:take/2 API

### CODE REVIEW

File: src/emqttd_session.erl [hunk 1]
```diff
@@ -336,10 +333,14 @@ prioritise_info(Msg, _Len, _State) ->
         _                -> 0
     end.
 
-handle_call({publish, Msg = #mqtt_message{qos = ?QOS_2, pktid = PacketId}},
-            _From, State = #state{awaiting_rel      = AwaitingRel,
-                                  await_rel_timer   = Timer,
-                                  await_rel_timeout = Timeout}) ->
+handle_pre_hibernate(State = #state{client_id = ClientId}) ->
+    io:format("Session(~s) will hibernate!~n", [ClientId]),
```
Comment (grutabow) [inline src/emqttd_session.erl]: Replace io:format() with lager.

File: src/emqttd_session.erl [hunk 2]
```diff
@@ -409,55 +410,59 @@ handle_cast({unsubscribe, _From, TopicTable},
...
-            noreply(State#state{awaiting_rel = AwaitingRel1});
-        error ->
-            ?LOG(warning, "Cannot find PUBREL: ~p", [PacketId], State),
-            emqttd_metrics:inc('packets/pubrel/missed'),
-            noreply(State)
-    end;
+    {noreply,
+     case maps:take(PacketId, AwaitingRel) of
+         {Msg, AwaitingRel1} ->
+             spawn(emqttd_server, publish, [Msg]), %%:)
```
Comment (grutabow) [inline src/emqttd_session.erl]: It is better to increase min_heap_size of the spawned process.

File: src/emqttd_client.erl [hunk 3]
```diff
@@ -94,71 +98,80 @@ session(CPid) ->
...
+    end.
+
+prioritise_call(Msg, _From, _Len, _State) ->
+    case Msg of info -> 10; stats -> 10; state -> 10; _ -> 5 end.
+
+prioritise_info(Msg, _Len, _State) ->
+    case Msg of {redeliver, _} -> 5; _ -> 0 end.
+
+handle_pre_hibernate(State = #client_state{peername = Peername}) ->
+    io:format("Client(~s) will hibernate!~n", [esockd_net:format(Peername)]),
```
Comment (emqplus) [inline src/emqttd_client.erl]: Ok, Delete the io:format line.

File: src/emqttd_session.erl (hunk 2 above)
Comment (emqplus) [inline src/emqttd_session.erl]: Increase the min_heap_size according to the Msg size?

Comment (grutabow) [inline src/emqttd_session.erl]: Measure the memory size emqttd_server:publish() may use. For example:
1) Set the max_heap_size to 1000, and kill = false, error_logger = true.
2) Run some cases and check the log that any max heap error is present. If present, it means max_heap_size is not big enough to held variables in these functions.
3) Increase max_heap_size to 2000.
4) Run same cases again, If no error, 2000 is a safe value.
5) min_heap_size should be at least 2000 within the spawn_opt.

File: src/emqttd_gc.erl [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+-module(emqttd_gc).
+
+-author("Feng Lee <feng@emqtt.io>").
+
+-export([conn_max_gc_count/0, reset_conn_gc_count/2, maybe_force_gc/2]).
+
+-spec(conn_max_gc_count() -> integer()).
+conn_max_gc_count() ->
+    case emqttd:env(conn_force_gc_count) of
+        {ok, I} when I > 0 -> I + rand:uniform(I);
```
Comment (grutabow) [inline src/emqttd_gc.erl]: Maybe this random value varies too much, and it is hard to compare performances of this parameter.

File: src/emqttd_client.erl [hunk 5]
```diff
@@ -48,36 +48,40 @@
...
 -define(SOCK_STATS, [recv_oct, recv_cnt, send_oct, send_cnt, send_pend]).
 
 -define(LOG(Level, Format, Args, State),
-            lager:Level("Client(~s): " ++ Format, [State#client_state.connname | Args])).
+            lager:Level("Client(~s): " ++ Format,
+                        [esockd_net:format(State#client_state.peername) | Args])).
 
 start_link(Conn, Env) ->
-    {ok, proc_lib:spawn_link(?MODULE, init, [[Conn, Env]])}.
+    {ok, proc_lib:spawn_opt(?MODULE, init, [[Conn, Env]], [link | ?FULLSWEEP_OPTS])}.
```
Comment (grutabow) [inline src/emqttd_client.erl]: Are you sure that fullsweep is good for this process?

**Pull request merged**